lib/srv/db/objects/importer.go

```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+func calculateUpdates(ctx context.Context, cfg Config, objects map[string]*objWithExpiry, objsNew map[string]*dbobjectv1.DatabaseObject) map[string]*objWithExpiry {
+	updated := make(map[string]*objWithExpiry)
+	now := cfg.Clock.Now()
+	expiry := now.Add(cfg.ObjectTTL)
+
+	var countNew, countChanged, countRefresh int
+
+	for key, objNew := range objsNew {
+		objOld, found := objects[key]
+		// completely new object
+		if !found {
+			countNew++
+			updated[key] = &objWithExpiry{
+				obj:    objNew,
+				expiry: expiry,
+			}
+			continue
+		}
+
+		// previously seen object, check for changes.
+		// we can safely use proto.Equal as these objects are free of revision etc.
+		if !proto.Equal(objOld.obj, objNew) {
+			countChanged++
+			updated[key] = &objWithExpiry{
+				obj:    objNew,
+				expiry: expiry,
+			}
+			continue
+		}
+
+		// do we need to refresh the object in backend?
+		if objOld.expiry.Sub(now) < cfg.RefreshThreshold {
+			countRefresh++
+			updated[key] = &objWithExpiry{
+				obj:    objNew,
+				expiry: expiry,
+			}
+		}
+	}
+
+	cfg.Log.DebugContext(ctx, "Objects to update", "new", countNew, "changed", countChanged, "refreshed", countRefresh)
+
+	return updated
+}
+
+func (i *singleDatabaseImporter) updateObjects(ctx context.Context, updated map[string]*objWithExpiry) {
+	client := i.cfg.AuthClient.DatabaseObjectsClient()
+
+	// upsert changed objects
+	var errs []error
+	for key, objNew := range updated {
+		i.objects[key] = objNew
+		clone := proto.Clone(objNew.obj).(*dbobjectv1.DatabaseObject)
+		clone.Metadata.Expires = timestamppb.New(objNew.expiry)
+		_, err := client.UpsertDatabaseObject(ctx, clone)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		i.cfg.Log.ErrorContext(ctx, "Errors occurred when updating objects", "error_count", len(errs), "errs", errs)
+	}
+}
+
+func (i *singleDatabaseImporter) fetchObjects(ctx context.Context) ([]*dbobjectv1.DatabaseObject, error) {
+	rules, err := i.cfg.AuthClient.GetDatabaseObjectImportRules(ctx)
+	if err != nil {
+		i.cfg.Log.ErrorContext(ctx, "Failed to fetch database object import rules", "err", err)
+		return nil, trace.Wrap(err)
+	}
+
+	dbNameFilter := databaseobjectimportrule.CalculateDatabaseNameFilter(rules, i.database)
+	objsFetched, err := i.fetcher.FetchDatabaseObjects(ctx, dbNameFilter)
+	if err != nil {
+		i.cfg.Log.WarnContext(ctx, "Error while fetching database objects", "err", err, "imported", len(objsFetched))
+	}
+
+	// TODO(Tener): switch to cfg.log logger once ApplyDatabaseObjectImportRules is on slog too.
+	objsImported, errCount := databaseobjectimportrule.ApplyDatabaseObjectImportRules(logrus.WithContext(ctx).WithField("db", i.database), rules, i.database, objsFetched)
+	counts, _ := permissions.CountObjectKinds(objsImported)
+	i.cfg.Log.InfoContext(ctx, "Database objects imported", "counts", counts, "err_count", errCount, "total", len(objsImported))
```

Comment:
Looking at this with fresh eyes I think some refactoring is due. This code piece is lifted from Postgres DAC implementation, but it would be good to abstract away this pattern (i.e. fetch rules, fetch objects, apply rules, return finished objects).

Reply:
I've done the refactoring I've mentioned above. All object-fetching code is using the same interface now, with import rules being applied consistently by a wrapping fetcher.